### PR TITLE
Factor out Host and Port Access Lists to a Common Bundle

### LINKED ISSUE #18438: Move AccessLists into independent bundle
Continuing with the theme of #18081 - the AccessList code currently in the channel framework, `com.ibm.ws.tcpchannel.internal.AccessLists` `com.ibm.ws.udpchannel.internal.AccessLists`, should be moved into a new common bundle since we'll continue to need this functionality from both netty and the legacy channel framework.

Additionally, most of the (TCP) `AccessLists` functionality is duplicated in the HTTP channel (`com.ibm.ws.transport.http`), under `com.ibm.ws.http.channel.internal.filter`. The new common AccessList code should be made generic enough to be reused by the HTTP, so that we can get rid of the duplicate code in `com.ibm.ws.http.channel.internal.filter`.

### PR DESCRIPTION
**Pull request by hutchig**
Signed-off-by: Gordon Hutchison <Gordon.Hutchison@gmail.com>
fixes #18438

### CODE REVIEW

**Comment by hutchig** [comment]
#build

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/tcpchannel/internal/TCPChannelConfiguration.java [hunk 1]
```diff
@@ -1258,4 +1260,52 @@ protected boolean getWaitToAccept() {
         return this.waitToAccept;
     }
 
+    /**
```
Comment (hutchig) [inline dev/com.ibm.ws.channelfw/src/com/ibm/ws/tcpchannel/internal/TCPChannelConfiguration.java:1263]: better comments

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/tcpchannel/internal/TCPChannelConfiguration.java [hunk 2]
```diff
@@ -1258,4 +1260,52 @@ protected boolean getWaitToAccept() {
...
+     * This class is used to protect consumers of the access list keys from having
+     * visibility of this class, it collects together the methods required by the
+     * common access lists code without that code having to know about this consuming
+     * type.
+     */
+    private class AccessListKeysProvider implements AccessListKeysFacade {
+
+        TCPChannelConfiguration delegate;
+
+        AccessListKeysProvider(TCPChannelConfiguration config) {
```
Comment (hutchig) [inline dev/com.ibm.ws.channelfw/src/com/ibm/ws/tcpchannel/internal/TCPChannelConfiguration.java:1280]: convert to generic and make UDP and TCP common?

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/udpchannel/internal/UDPChannelConfiguration.java [hunk 3]
```diff
@@ -359,4 +361,37 @@ public void setChannelReceiveBufferSize(int size) {
...
+
+    /**
+     * This class is used to protect consumers of the access list keys from having
+     * visibility of this class, it collects together the methods required by the
+     * common access lists code without that code having to know about this consuming
+     * type.
+     */
+    public class AccessListKeysProvider implements AccessListKeysFacade {
+
+        UDPChannelConfiguration delegate;
```
Comment (hutchig) [inline dev/com.ibm.ws.channelfw/src/com/ibm/ws/udpchannel/internal/UDPChannelConfiguration.java:379]: make generic and common with TCP?

File: dev/com.ibm.websphere.appserver.features/visibility/private/com.ibm.websphere.appserver.channelfw-1.0.feature [hunk 4]
```diff
@@ -8,7 +8,9 @@ IBM-Process-Types: server, \
  com.ibm.ws.channelfw, \
  com.ibm.ws.timer, \
  io.openliberty.endpoint, \
- com.ibm.ws.wsbytebuffer
+ com.ibm.ws.wsbytebuffer, \
+ io.openliberty.accesslists.internal
```
Comment (hutchig) [inline dev/com.ibm.websphere.appserver.features/visibility/private/com.ibm.websphere.appserver.channelfw-1.0.feature:12]: Is it OK to have '.internal' at the end of the project name?

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/udpchannel/internal/UDPChannel.java [hunk 5]
```diff
@@ -43,13 +43,13 @@
     private DiscriminationProcess discriminationProcess = null;
     private boolean stoppedFlag = true;
     private String displayableHostName = null;
-    private LinkedList<UDPConnLink> inUse = new LinkedList<UDPConnLink>();
+    private final LinkedList<UDPConnLink> inUse = new LinkedList<UDPConnLink>();
```
Comment (hutchig) [inline dev/com.ibm.ws.channelfw/src/com/ibm/ws/udpchannel/internal/UDPChannel.java]: Eclipse added this 'final',  I will remove it.

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/udpchannel/internal/UDPChannel.java [hunk 6]
```diff
@@ -73,21 +73,24 @@ public UDPChannel(UDPChannelConfiguration config, WorkQueueManager workQueueMana
     /*
      * @see com.ibm.wsspi.channelfw.OutboundChannel#getDeviceAddress()
      */
+    @Override
```
Comment (hutchig) [inline dev/com.ibm.ws.channelfw/src/com/ibm/ws/udpchannel/internal/UDPChannel.java]: Eclipse added these, '"Override' s I will remove them

File: dev/com.ibm.ws.channelfw/src/com/ibm/ws/udpchannel/internal/UDPChannelConfiguration.java [hunk 7]
```diff
@@ -359,4 +361,37 @@ public void setChannelReceiveBufferSize(int size) {
         return this.addressIncludeList;
     }
 
+    /**
```
Comment (hutchig) [inline dev/com.ibm.ws.channelfw/src/com/ibm/ws/udpchannel/internal/UDPChannelConfiguration.java:364]: Better comment

File: dev/io.openliberty.accesslists.internal/bnd.bnd [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+Export-Package: \
+	io.openliberty.accesslists;version="1.0";provide:=true, \
+	com.ibm.ws.tcpchannel.internal;-split-package:=merge-last;version="1.0";provide:=true, \
+	com.ibm.ws.udpchannel.internal;-split-package:=merge-last;version="1.0";provide:=true
+
+Import-Package: \
+	com.ibm.ws.bytebuffer.internal, \
+	${defaultPackageImport}
+
+Private-Package: \
```
Comment (hutchig) [inline dev/io.openliberty.accesslists.internal/bnd.bnd]: I will remove the Private-Package as there are no resources folders

Comment (hutchig) [reply dev/io.openliberty.accesslists.internal/bnd.bnd]: String ADDRESS_EXCLUDE_LIST_INVALID = "ADDRESS_EXCLUDE_LIST_INVALID";
    /** Config has an invalid IP include list */
    String ADDRESS_INCLUDE_LIST_INVALID = "ADDRESS_INCLUDE_LIST_INVALID";
Continue to be used in the channelfw project in checkAndSetValues,

Incidentally, I do no think
    String HOST_NAME_EXCLUDE_LIST_INVALID = "HOST_NAME_EXCLUDE_LIST_INVALID";
    /** Config has an invalid hostname include list */
    String HOST_NAME_INCLUDE_LIST_INVALID = "HOST_NAME_INCLUDE_LIST_INVALID";
are used anywhere

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#run-libby-bot

File: dev/io.openliberty.accesslists.internal/bnd.bnd [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+and HTTP channels version ${bVersion}
+
+WS-TraceGroup: ChannelFramework
+
+# For each exported package, create (in that package) a package-info.java
+# file, and place an @version javadoc tag in package-level javadoc. 
+# Append ";provide:=true" if this bundle also provides an implementation
+# for the exported API.
+Export-Package: \
+	io.openliberty.accesslists;version="1.0";provide:=true, \
```
Comment (wtlucy) [inline dev/io.openliberty.accesslists.internal/bnd.bnd]: is there a reason for exporting the package version here rather than in a `package-info.java`?

Comment (hutchig) [reply dev/io.openliberty.accesslists.internal/bnd.bnd]: No, just a copy/paste error. I will fix.

Comment (hutchig) [reply dev/io.openliberty.accesslists.internal/bnd.bnd]: done.

**Comment by hutchig** [comment]
#build

**Review by wtlucy — COMMENTED** [review]
The changes you have here look good to me. But as we've discussed on Slack, even though you haven't changed any implementation details here I'm concerned that there aren't any existing tests for this functionality. As an added bonus here it would be great if you could add some test cases. A few unit tests in the new `io.openliberty.accesslists.internal` project would be helpful for validating the basic functionality, but even more helpful in the context of this PR would be to add a more end-to-end FAT case which verifies that server.xml configs are actually applied correctly:
```
<httpEndpoint ...> 
<tcpOptions addressExcludeList="..." addressIncludeList="..." hostNameExcludeList="..." hostNameIncludeList="..."/> 
</httpEndpoint>
``` 
Unfortunately there isn't a great FAT bucket for such a test in this repo. `com.ibm.ws.transport.http_test` in WS-CD-Open would make sense, but I'd prefer not to mix repos for this. The best option might be to create a _new_ FAT project like `io.openliberty.transport.http_fat`, which we'd be able to reuse for future cases like this.

**Comment by mrsaldana** [comment]
I agree that moving forward we should have better validation of these implementations. I was looking over the existing base channel framework test/bvt/fat and there does not seem to be much coverage. Adding FAT coverage and unit testing, where it makes sense, will add a lot of value.

**Comment by hutchig** [comment]
For test inspiration see also:
   https://github.com/OpenLiberty/open-liberty/pull/14661/commits
   https://github.ibm.com/websphere/WS-CD-Open/pull/23081/files

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#run_libby_bot

**Comment by hutchig** [comment]
#libbybot

File: dev/io.openliberty.transport.http_fat/publish/files/.gitignore [hunk 10]
```diff
@@ -0,0 +1 @@
+/CookieDate.zip
```
Comment (wtlucy) [inline dev/io.openliberty.transport.http_fat/publish/files/.gitignore]: what is this referencing?

File: dev/io.openliberty.transport.http_fat/publish/files/AccessListsTests_testA10_server.xml [hunk 11]
```diff
@@ -0,0 +1,24 @@
+<!-- Copyright (c) 2017 IBM Corporation and others. All rights reserved.
```
Comment (wtlucy) [inline dev/io.openliberty.transport.http_fat/publish/files/AccessListsTests_testA10_server.xml]: can you fix the copyright date for all of these server.xmls, and remove the `CLIENT_HOST CLIENT_ADDR CLIENT_PORT` from these files (unless I'm missing something and the string is used somehow)?

File: dev/io.openliberty.transport.http_fat/publish/files/.gitignore (hunk 10 above)
Comment (hutchig) [reply dev/io.openliberty.transport.http_fat/publish/files/.gitignore]: I will investigate!

File: dev/io.openliberty.transport.http_fat/publish/files/AccessListsTests_testA10_server.xml (hunk 11 above)
Comment (hutchig) [reply dev/io.openliberty.transport.http_fat/publish/files/AccessListsTests_testA10_server.xml]: Ok will do.

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

File: dev/io.openliberty.transport.http_fat/publish/files/AccessListsTests_testA10_server.xml (hunk 11 above)
Comment (hutchig) [reply dev/io.openliberty.transport.http_fat/publish/files/AccessListsTests_testA10_server.xml]: Done.

File: dev/io.openliberty.transport.http_fat/publish/files/.gitignore (hunk 10 above)
Comment (hutchig) [reply dev/io.openliberty.transport.http_fat/publish/files/.gitignore]: It seems to be common with only the project  I copy and pasted as a start line which is curious - the http2_fat project: 
image

Comment (hutchig) [reply dev/io.openliberty.transport.http_fat/publish/files/.gitignore]: Gone.

Comment (wtlucy) [reply dev/io.openliberty.transport.http_fat/publish/files/.gitignore]: I see - looks like some cleanup is needed in the http2_fat as well. Thanks for finding this!

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
For the intermittant failures, the extra debug shows that, for example: 
[ https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_cX1tgEL8Eeyda8UYcX_yfQ&tab=com.ibm.team.build.web.ui.internal.editors.result.overview.OverviewPage ]
text fed to jvm File write is "
```
<server description="Generation date: Thu Nov 11 16:57:24 PST 2021">
    <featureManager>
        <feature>componenttest-1.0</feature>
        <feature>servlet-4.0</feature>
        <feature>timedexit-1.0</feature>
    </featureManager>
    <httpEndpoint host="*" httpPort="${bvt.prop.HTTP_default}" httpsPort="${bvt.prop.HTTP_default.secure}" id="defaultHttpEndpoint">
        <tcpOptions addressExcludeList="0:0:0:0:0:0:0:1, 0:0:0:0:007F:1:1:1, 007F:0:0:1:0:0:0:0"/>
    </httpEndpoint>
    <application location="app1.war"/>
</server>
```
but the code:
```
debug("Unexpected server access problem for " + reqURI + "server config: " + server.getServerConfiguration());

[file:///Users/hutchig/Downloads/fatoutput-io.openliberty.transport.http_fat-3/results/output.txt]
Outputs:
[11/11/2021 16:57:45:438 PST] 001 Utils                          debug                          I Unexpected server access problem for /app1/Backendserver config: ServerConfiguration
FeatureManager{feature="componenttest-1.0" feature="servlet-4.0" feature="timedexit-1.0" }[HttpEndpoint{id="defaultHttpEndpoint" host="*" httpPort="${bvt.prop.HTTP_default}" httpsPort="${bvt.prop.HTTP_default.secure}" TcpOptions{id="null" }}][Application{location="app1.war" }]
```
Note null for TcpOptions.

[Update: I have improved the FAT infrastructure TcpOptions class to improve toString and used it in the test scaffolding.
Both tests that have failed usually pass - so it may be a timing issue with server config update...investigating.

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
I have requested multiple builds as I have put more FFDC in for an intermittent problem. (Also the build is locked currently - so this is not likely to materially impact others at the moment).

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
build

**Comment by hutchig** [comment]
build

**Comment by hutchig** [comment]
build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#build

File: dev/io.openliberty.transport.http_fat/bnd.bnd [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+# Define additional tested features that are NOT present in any XML files in this bucket.
+# In this case, Java EE 8 features are added programmatically at runtime by the RepeatTests rule.
+tested.features:\
+	servlet-4.0, \
+	servlet-5.0
+
+-buildpath:\
+	io.openliberty.org.apache.commons.logging;version=latest, \
+	io.openliberty.org.apache.commons.codec;version=latest, \
+	com.ibm.ws.http2.client;version=latest, \
```
Comment (wtlucy) [inline dev/io.openliberty.transport.http_fat/bnd.bnd]: can you remove this `com.ibm.ws.http2.client`?

Comment (hutchig) [reply dev/io.openliberty.transport.http_fat/bnd.bnd]: Yes.

File: dev/io.openliberty.transport.http_fat/fat/src/io/openliberty/transport/http_fat/accesslists/Utils.java [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+            final StringBuilder sb = new StringBuilder(custom);
+            // Substitute any keys for values from Properties - short and simple (prefixes etc. not needed)
+            if (p != null) {
+                p.forEach((k, v) -> sb.replace(0, sb.length(), sb.toString().replaceAll((String) k, (String) v)));
+            }
+            custom = sb.toString();
+            return custom;
+        } catch (IOException e) {
+            e.printStackTrace();
+            Assert.fail("failed to cusotmize " + src + " from " + p + " due to " + e.getMessage());
```
Comment (wtlucy) [inline dev/io.openliberty.transport.http_fat/fat/src/io/openliberty/transport/http_fat/accesslists/Utils.java]: extremely minor: `cusotmize` -> `customize`

**Review by wtlucy — APPROVED** [review]
looks great to me, thanks for the work here

File: dev/io.openliberty.transport.http_fat/fat/src/io/openliberty/transport/http_fat/accesslists/Utils.java (hunk 13 above)
Comment (hutchig) [reply dev/io.openliberty.transport.http_fat/fat/src/io/openliberty/transport/http_fat/accesslists/Utils.java]: Apologies and thanks for spotting it.

File: dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java [hunk 14]
```diff
@@ -5159,7 +5159,9 @@ public void updateLogOffset(String logFile, Long newLogOffset) {
                             String regexp = it.next();
                             if (Pattern.compile(regexp).matcher(line).find()) {
                                 it.remove();
-                                break;
+                                //There used to be a break here but if a user passed in a
```
Comment (hutchig) [inline dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java:5162]: I feel guilty walking past the egregious pattern recompile for every line. I shudder to think how many times the smarter planet regex is compiled for each build.

**Comment by hutchig** [comment]
#build

**Comment by hutchig** [comment]
#libbybot

File: dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java (hunk 14 above)
Comment (hutchig) [reply dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java:5162]: Raised #19426

Comment (wtlucy) [reply dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java:5162]: That's a good point, caching even just the basic smarter planet regex would probably be a good improvement

**Comment by hutchig** [comment]
Checked that all related http_fat and transport tests were OK - so will merge.

**Pull request merged**